### ISSUE TITLE
Error while deploying azure-iot-operations failed to create Processor with error " The resource write operation failed to complete successfully, because it reached terminal provisioning state 'Failed'.

### ISSUE THREAD

**Issue by ashwinimhaiskar** [Azure/azure-iot-operations#29]
I was trying to deploy Azure IOT operations from Azure portal.
Deployment failed with below error.

ResourceDeploymentFailure","target":"/subscriptions/****/resourceGroups/*****/providers/Microsoft.Kubernetes/connectedClusters/*******/providers/Microsoft.KubernetesConfiguration/extensions/processor","message":"The resource write operation failed to complete successfully, because it reached terminal provisioning state 'Failed'.","details":[***"code":"ExtensionOperationFailed","message":"The extension operation failed with the following error:  Error: [ **InnerError: [Helm installation failed : Timed out waiting for the resource to come to a ready/completed state : Recommendation Please contact Microsoft support for further inquiries : InnerError [release processor failed, and has been uninstalled due to atomic being set: timed out waiting for the condition]]] occurred while doing the operation : [Create] on the config, For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG, For more application specific troubleshooting visit: please reach out to Microsoft tech care if you need help with debugging the problems with your extension** resource.","additionalInfo":[]***]***]***

How to solve this issue ?

**Comment by chgennar** [comment]
Hi @ashwinimhaiskar, 

Was step 9 followed from the documentation here?
https://learn.microsoft.com/en-us/azure/iot-operations/get-started/quickstart-deploy?tabs=windows#deploy-azure-iot-operations 

The `az iot ops init` command from this step needs to be run on the cluster where AIO will be deployed to initialize some dependencies.

**Comment by ashwinimhaiskar** [comment]
Yes. this step was followed. We also tried to deploy application from forked code by changing required details through git hub directly.
Still, we are getting same issue.
In Error details some generic message is coming so not able to understand what is missing.
Coping Error message below. 

{
    "status": "Failed",
    "error": {
        "code": "ExtensionOperationFailed",
        "message": "The extension operation failed with the following error:  Error: [ InnerError: [Helm installation failed : Timed out waiting for the resource to come to a ready/completed state : Recommendation Please contact Microsoft support for further inquiries : InnerError [release processor failed, and has been uninstalled due to atomic being set: timed out waiting for the condition]]] occurred while doing the operation : [Create] on the config, For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG, For more application specific troubleshooting visit: please reach out to Microsoft tech care if you need help with debugging the problems with your extension resource.",
        "additionalInfo": []
    }
}

**Comment by ashwinimhaiskar** [comment]
Yes. this step was followed. We also tried to deploy application from forked code by changing required details through git hub directly.
Still, we are getting same issue.
In Error details some generic message is coming so not able to understand what is missing.
Coping Error message below. 

{
    "status": "Failed",
    "error": {
        "code": "ExtensionOperationFailed",
        "message": "The extension operation failed with the following error:  Error: [ InnerError: [Helm installation failed : Timed out waiting for the resource to come to a ready/completed state : Recommendation Please contact Microsoft support for further inquiries : InnerError [release processor failed, and has been uninstalled due to atomic being set: timed out waiting for the condition]]] occurred while doing the operation : [Create] on the config, For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG, For more application specific troubleshooting visit: please reach out to Microsoft tech care if you need help with debugging the problems with your extension resource.",
        "additionalInfo": []
    }
}

**Comment by ruonke** [comment]
@ashwinimhaiskar  After you run init, check your kube log for errors:

 kubectl get events -n azure-iot-operations -o json

You should find something like MountVolume.SetUp failed 403 error with 
message :  Caller is not authorized to perform action on resource. 
Caller: appid=<this is the id you need>

Go to azure portal, find the keyvault you used in the init command, select secrets, azure-iot-operations and then Access Control (IAM)
Grant Key Vault Secrets Officer right to the appid above and run the init command with the appid parameter.


The instructions on the linked microsoft page are kinda bad, but when you run the init for first time with the --no-deploy parameter, the output says:
   KeyVault CSI Driver 
     - Use app '<your appid>'
     - Ensure KeyVault access policy
     - Ensure secret name 'azure-iot-operations' for service account
     - Deploy driver to cluster
     - Configure driver 

Once you run the init with --no-deploy, it will generate your appid, the App (id) is for --sp-app-id parameter
Now that you have the appid, you can configure the keyvault access policy via Azure Portal(as explained above)
When you run the init again to deploy, you need to omit the keyvault definition and use the sp-app-id parameter.

**Comment by BillmanH** [comment]
Same issue exactly. All of the other components have installed. 

My error in the deployment log:
{\"status\":\"Failed\",\"error\":{\"code\":\"ResourceOperationFailure\",\"message\":\"The resource operation completed with terminal provisioning state 'Failed'.\",\"details\":[{\"code\":\"ExtensionOperationFailed\",\"message\":\"The extension operation failed with the following error:  Helm Upgrade Failed : Timed out waiting for the resource to come to a ready/completed state : Recommendation Please contact Microsoft support for further inquiries : InnerError [ InnerError: [Helm installation failed : Timed out waiting for the resource to come to a ready/completed stateStatefulSet is not ready: azure-iot-operations/aio-dp-reader-worker. 0 out of 1 expected pods are ready : Recommendation Please contact Microsoft support for further inquiries : InnerError [release processor failed, and has been uninstalled due to atomic being set: timed out waiting for the condition]]], For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG, For more application specific troubleshooting visit: please reach out to Microsoft tech care if you need help with debugging the problems with your extension resource.\",\"additionalInfo\":[]}]}}

I have run the steps above, including keyvault access. You can tell that this has been completed because the other components are installed. 
![image](https://github.com/Azure/azure-iot-operations/assets/10108711/47ba479b-eeba-4a05-b0c3-0bc260893493)
Only the `processor` component did not install. 

On running: 
`kubectl get events -n azure-iot-operations -o json`

I get the response: 
```
{
    "apiVersion": "v1",
    "items": [],
    "kind": "List",
    "metadata": {
        "resourceVersion": ""
    }
}
```

**Comment by ashwinimhaiskar** [comment]
@ruonke 
azure IOT operations were deployed properly and started working for us on 22nd dec till 28th Dec ,on 29th morning cluster was offline and not connected to Azure ,so i deleted everything and tried deploying it again but till now it's not happening .

I followed above steps also which you mentioned but still processor deployment is failing .

On running:
kubectl get events -n azure-iot-operations -o json

In logs i am getting below details. 
{
  "apiVersion": "v1",
  "count": 201,
  "eventTime": null,
  "firstTimestamp": "2024-01-05T05:22:38Z",
  "involvedObject": {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "name": "runner-local-aio-dp-runner-worker-0",
    "namespace": "azure-iot-operations",
    "resourceVersion": "9357",
    "uid": "55288e01-05ef-4d30-bb01-bc1be14dd94b"
  },
  "kind": "Event",
  "lastTimestamp": "2024-01-05T06:12:39Z",
  "message": "no persistent volumes available for this claim and no storage class is set",
  "metadata": {
    "creationTimestamp": "2024-01-05T05:22:39Z",
    "name": "runner-local-aio-dp-runner-worker-0.17a75c04a29524d6",
    "namespace": "azure-iot-operations",
    "resourceVersion": "591259",
    "uid": "ddd9efcb-6565-4dd9-b5ba-6d62e411c170"
  },
  "reason": "FailedBinding",
  "reportingComponent": "",
  "reportingInstance": "",
  "source": {
    "component": "persistentvolume-controller"
  },
  "type": "Normal"
}
Error on azure portal after deployment failed .
{
    "status": "Failed",
    "error": {
        "code": "ExtensionOperationFailed",
        "message": "The extension operation failed with the following error:  Error: [ InnerError: [Helm installation failed : Timed out waiting for the resource to come to a ready/completed state : Recommendation Please contact Microsoft support for further inquiries : InnerError [release processor failed, and has been uninstalled due to atomic being set: timed out waiting for the condition]]] occurred while doing the operation : [Create] on the config, For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG, For more application specific troubleshooting visit: please reach out to Microsoft tech care if you need help with debugging the problems with your extension resource.",
        "additionalInfo": []
    }
}
Also attached Deployment failed  image 

![image](https://github.com/Azure/azure-iot-operations/assets/37296095/cca58ce8-cdeb-46f7-a5b9-c84b6fa97db0)

**Comment by Mzamankhan** [comment]
@ashwinimhaiskar Are you still blocked on this issue?

Can you please provide a full support bundle ( https://learn.microsoft.com/en-us/cli/azure/iot/ops?view=azure-cli-latest ).

**Comment by digimaun** [comment]
Hi @ashwinimhaiskar and @BillmanH are you deploying on an Ubuntu environment? If so do you have nfs-common installed as outlined in the [cluster setup](https://learn.microsoft.com/en-us/azure/iot-operations/deploy-iot-ops/howto-prepare-cluster?tabs=ubuntu#create-a-cluster) section?

What is the permission model of the keyvault you are using for the CSI Driver setup step? It must be set to "Vault Access Policy" and **not** Azure RBAC.
![image](https://github.com/Azure/azure-iot-operations/assets/28658112/c2905aa3-bf7e-4268-ae3f-eae23a3258e4)

**Comment by BillmanH** [comment]
@digimaun , yes that seems to have done it. I must have missed this in the setup docs. Thanks!

**Comment by digimaun** [comment]
Excellent @BillmanH , glad to hear and thank you for reporting back on what worked.

**Comment by ashwinimhaiskar** [comment]
@digimaun   Yesterday during IST time Zone. I tried to deploy once again, and it got deployed without doing any changes.

**Comment by mahomedalid** [comment]
I am having this issue today. Tried deploying with bicep into westus2, and using the cli az iot ops init, same issue.

The data processor part fails:

![image](https://github.com/Azure/azure-iot-operations/assets/952392/fcc59664-608d-47c2-a7d8-279910b18ae9)

This is when trying with bicep:
```

{
  "status": "Failed",
  "error": {
    "code": "DeploymentFailed",
    "target": "/subscriptions/a6cdfc52-917e-4744-8785-031f64901c9b/resourceGroups/rg-aio/providers/Microsoft.Resources/deployments/aio-deployment-aio-deployment-1641",
    "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.",
    "details": [
      {
        "code": "ResourceDeploymentFailure",
        "target": "/subscriptions/a6cdfc52-917e-4744-8785-031f64901c9b/resourceGroups/rg-aio/providers/Microsoft.Resources/deployments/aio-rg-aio-20240109222244",
        "message": "The resource write operation failed to complete successfully, because it reached terminal provisioning state Failed.",
        "details": [
          {
            "code": "DeploymentFailed",
            "target": "/subscriptions/a6cdfc52-917e-4744-8785-031f64901c9b/resourceGroups/rg-aio/providers/Microsoft.Resources/deployments/aio-rg-aio-20240109222244",
            "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.",
            "details": [
              {
                "code": "ResourceDeploymentFailure",
                "target": "/subscriptions/a6cdfc52-917e-4744-8785-031f64901c9b/resourceGroups/rg-aio/providers/Microsoft.Kubernetes/connectedClusters/cluster-408/providers/Microsoft.KubernetesConfiguration/extensions/azure-iot-operations",
                "message": "The resource write operation failed to complete successfully, because it reached terminal provisioning state Failed.",
                "details": [
                  {
                    "code": "ExtensionOperationFailed",
                    "message": "The extension operation failed with the following error:  Failed to resolve the extension version from the given values. Please refer https://aka.ms/k8s-extension-type-versions to find the correct version for your installation, For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG.",
                    "additionalInfo": []
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
}
```

**Comment by digimaun** [comment]
@mahomedalid have tried the suggestions here https://github.com/Azure/azure-iot-operations/issues/29#issuecomment-1881703064 ?

**Comment by meenag16** [comment]
@mahomedalid following up, did the suggestions @digimaun mentioned work out?

**Comment by ashwinimhaiskar** [comment]
> @ruonke azure IOT operations were deployed properly and started working for us on 22nd dec till 28th Dec ,on 29th morning cluster was offline and not connected to Azure ,so i deleted everything and tried deploying it again but till now it's not happening .
> 
> I followed above steps also which you mentioned but still processor deployment is failing .
> 
> On running: kubectl get events -n azure-iot-operations -o json
> 
> In logs i am getting below details. { "apiVersion": "v1", "count": 201, "eventTime": null, "firstTimestamp": "2024-01-05T05:22:38Z", "involvedObject": { "apiVersion": "v1", "kind": "PersistentVolumeClaim", "name": "runner-local-aio-dp-runner-worker-0", "namespace": "azure-iot-operations", "resourceVersion": "9357", "uid": "55288e01-05ef-4d30-bb01-bc1be14dd94b" }, "kind": "Event", "lastTimestamp": "2024-01-05T06:12:39Z", "message": "no persistent volumes available for this claim and no storage class is set", "metadata": { "creationTimestamp": "2024-01-05T05:22:39Z", "name": "runner-local-aio-dp-runner-worker-0.17a75c04a29524d6", "namespace": "azure-iot-operations", "resourceVersion": "591259", "uid": "ddd9efcb-6565-4dd9-b5ba-6d62e411c170" }, "reason": "FailedBinding", "reportingComponent": "", "reportingInstance": "", "source": { "component": "persistentvolume-controller" }, "type": "Normal" } Error on azure portal after deployment failed . { "status": "Failed", "error": { "code": "ExtensionOperationFailed", "message": "The extension operation failed with the following error: Error: [ InnerError: [Helm installation failed : Timed out waiting for the resource to come to a ready/completed state : Recommendation Please contact Microsoft support for further inquiries : InnerError [release processor failed, and has been uninstalled due to atomic being set: timed out waiting for the condition]]] occurred while doing the operation : [Create] on the config, For general troubleshooting visit: https://aka.ms/k8s-extensions-TSG, For more application specific troubleshooting visit: please reach out to Microsoft tech care if you need help with debugging the problems with your extension resource.", "additionalInfo": [] } } Also attached Deployment failed image
> 
> ![image](https://private-user-images.githubusercontent.com/37296095/294408464-cca58ce8-cdeb-46f7-a5b9-c84b6fa97db0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwMDI2ODcsIm5iZiI6MTcwNjAwMjM4NywicGF0aCI6Ii8zNzI5NjA5NS8yOTQ0MDg0NjQtY2NhNThjZTgtY2RlYi00NmY3LWE1YjktYzg0YjZmYTk3ZGIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIzVDA5MzMwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExNjc1MGVmMmVlM2Y4YTIwZjlmY2VhOTlkZDZmOTI0MDAwMGVmNjU5ZmQ2YTE3ZDUxYWE3YTdhOTFhYWU5YjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gwd4blRK78SVWG1QBkyeCH098pSih6RDss_H3rIuoOk)

@ruonke  I have started facing same issue again ..

**Comment by ashwinimhaiskar** [comment]
@ruonke  We are still facing issue in deployment. we are completely stuck for deployment. 
Could please let me know what  could be the reason for above issue .
I am attaching Logs of  kubectl get events -n azure-iot-operations -o json
[event.json](https://github.com/Azure/azure-iot-operations/files/14108750/event.json)

**Comment by ruonke** [comment]
@ashwinimhaiskar Are you deploying to a host with enough resources? 
This sounds much like the attempts I had when I was installing to a host with 8Gb ram, I had to increase the VM size to 16Gb to get the system to install reliably.
It seems to run with 8Gb, using about 5-6Gb after you get the install process to complete, I've encountered few out-of-memory instances but I have other application pods deployed to the same instance which are using resources as well.

You mentioned that you removed the deployment and ran it again, at expected clean state, check 
kubectl get pv
and
kubectl get pvc -n azure-iot-operations
to see if the presistent volumes and the claims are actually removed.
The failure in logs sounds like there is some remaining volume or claim to one.

**Comment by ashwinimhaiskar** [comment]
@ruonke 
We are using below configuration VM Standard D4s v3 (4 vcpus, 16 GiB memory) with Windows (Windows Server 2022 Datacenter Azure Edition).I deleted VM and recreated and tried deploying IOT operations still I am getting same issue.
 kubectl get pvc  after running this command I am getting below issue.
No resources found.

 kubectl get pvc -n azure-iot-operations
NAME                                        STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
aio-dp-msg-store-js-aio-dp-msg-store-0      Pending                                                     3d2h
data-aio-dp-nfs-server-provisioner-0        Pending                                                     3d2h
nfs-provisioner                             Pending                                      nfs            12m
refdatastore-local-aio-dp-refdata-store-0   Pending                                                     3d2h
runner-local-aio-dp-runner-worker-0         Pending                                                     3d2h

**Comment by ruonke** [comment]
In our tests WSL hasn't been a platform I'd recommend for this. 
We managed to get iot ops running for one person out of four who tried installing it.
The setup works much better on ubuntu host.

**Comment by BillmanH** [comment]
Would it be possible (as it's in Preview and we know it's somewhat limited on OD). Could you just tell us the VM to use in the docs? You already have great docs on the memory limitations for the VM. But it might be easier to say "This SKU. This OS. Nothing else on it. This is the VM that we are testing on"
